geometry/parse-ply.cpp

```diff
@@ -0,0 +1,316 @@
+    parseHeader(std::ifstream& f)
+{
+    // check magic number
+    std::string line; 
+    nextLine(f, line);
+    if (line != "ply")
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Unexpected magic number of PLY file: " << line;
+    }
+
+    // check line specifying PLY format ("format ascii 1.0")
+    checkFormatLine(f); 
+    
+    // parse elements & comments
+    std::vector<PlyElement> elements;
+    std::vector<std::string> comments;
+    nextLine(f, line);
+    do
+    {
+        if (parseElement(f, line, elements, comments))
+        {
+            continue; // parseElement() jumps to the line after element
+                      // definition
+        }
+        else if (!parseComment(line, comments))
+        {
+            LOGTHROW(err4, std::runtime_error)
+                << "Encountered unexpected line in header (not a comment nor "
+                   "element definition): "
+                << line;
+        }
+
+        nextLine(f, line);
+    }
+    while (!ba::starts_with(line, "end_header"));
+
+    return std::make_pair(std::move(elements), std::move(comments));
+}
+
+
+void parsePly(PlyParserBase& parser, const fs::path& path)
+{
+    LOG(info1) << "Loading PLY file from " << path << ".";
+    std::ifstream f(path.native());
+    if (!f.good()) {
+        LOGTHROW(err4, std::runtime_error)
+            << "Unable to open file " << path << ".";
+    }
+    // not setting failbit - is set on eof
+    f.exceptions(std::ios::badbit);
+
+    // load header
+    std::vector<PlyElement> elements;
+    std::vector<std::string> comments;
+    std::tie(elements, comments) = parseHeader(f);
+    parser.loadHeader(elements, comments);
+    
+    // get face/vertex range
+    std::size_t fStart = 0;
+    std::size_t fEnd = 0;
+    std::size_t vStart = 0;
+    std::size_t vEnd = 0;
+    
+    std::size_t idx = 0;
+    for (const auto& el : elements)
+    {
+        if (el.name == "face")
+        {
+            if (fEnd != 0)
+            {
+                LOGTHROW(err4, std::runtime_error)
+                    << "Multiple elements named \"face\" found in header.";
+            }
+            fStart = idx;
+            fEnd = idx + el.count;
+        }
+        else if (el.name == "vertex")
+        {
+            if (vEnd != 0)
+            {
+                LOGTHROW(err4, std::runtime_error)
+                    << "Multiple elements named \"vertex\" found in header.";
+            }
+            vStart = idx;
+            vEnd = idx + el.count;
+        }
+        idx += el.count;
+    }
+    std::size_t totNum = idx;  // expected number of lines
+
+    LOG(info1) << "Expecting vertices on lines " << vStart << " - " << vEnd
+               << "; faces on lines " << fStart << " - " << fEnd;
+
+    // read the rest of the file
+    idx = 0;
+    std::size_t fRead = 0;
+    std::size_t vRead = 0;
+    while (!f.eof())
+    {
+        f.peek(); // skip last line at EOF
+        if (f.eof()) { break; }
```

Comment:
I think we know from the header which elements and how many data points we should parse. Maybe it's not necessary to check EOF/peek/etc and just set proper stream exception bit which will throw if we get an unexpected EOF.

Reply:
The idea here was to make sure `parser` reads the whole line. But yeah, the proposed solution is easier, I'll change it.

Reply:
It's dropped